packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/visibility/ModelsTreeVisibilityHandler.ts

```diff
@@ -0,0 +1,351 @@
+import type {
+  BaseIdsCache,
+  BaseTreeVisibilityHandlerOverrides,
+  TreeSpecificVisibilityHandler,
+} from "../../../common/internal/visibility/BaseVisibilityHelper.js";
+import type {
+  HierarchyVisibilityHandlerOverridableMethod,
+  HierarchyVisibilityOverrideHandler,
+  VisibilityStatus,
+} from "../../../common/UseHierarchyVisibility.js";
+import type { ModelsTreeIdsCache } from "../ModelsTreeIdsCache.js";
+import type { ModelsTreeFilterTargets } from "./FilteredTree.js";
+
+/**
+ * Functionality of Models tree visibility handler that can be overridden.
+ * Each callback is provided original implementation and reference to a `HierarchyVisibilityHandler`.
+ * @beta
+ */
+export interface ModelsTreeVisibilityHandlerOverrides extends BaseTreeVisibilityHandlerOverrides {
+  getSubjectsVisibilityStatus?: HierarchyVisibilityHandlerOverridableMethod<(props: { subjectIds: Id64Arg }) => Promise<VisibilityStatus>>;
+  getElementGroupingNodeVisibilityStatus?: HierarchyVisibilityHandlerOverridableMethod<(props: { node: GroupingHierarchyNode }) => Promise<VisibilityStatus>>;
+
+  changeSubjectsVisibilityStatus?: HierarchyVisibilityHandlerOverridableMethod<(props: { subjectIds: Id64Arg; on: boolean }) => Promise<void>>;
+  changeElementGroupingNodeVisibilityStatus?: HierarchyVisibilityHandlerOverridableMethod<
+    (props: { node: GroupingHierarchyNode; on: boolean }) => Promise<void>
+  >;
+}
+
+/** @internal */
+export interface ModelsTreeVisibilityHandlerProps {
+  idsCache: ModelsTreeIdsCache;
+  viewport: Viewport;
+  alwaysAndNeverDrawnElementInfo: AlwaysAndNeverDrawnElementInfo;
+  overrideHandler: HierarchyVisibilityOverrideHandler;
+  overrides?: ModelsTreeVisibilityHandlerOverrides;
+}
+
+/**
+ * Handles visibility status of models tree nodes.
+ *
+ * It knows how to get and change visibility status of nodes created by hierarchy definition.
+ * @internal
+ */
+export class ModelsTreeVisibilityHandler implements Disposable, TreeSpecificVisibilityHandler<ModelsTreeFilterTargets> {
+  private _visibilityHelper: ModelsTreeVisibilityHelper;
+
+  constructor(private readonly _props: ModelsTreeVisibilityHandlerProps) {
+    const baseIdsCache: BaseIdsCache = {
+      getCategories: (props) => this.getCategories(props),
+      getElementsCount: (props) => this.getElementsCount(props),
+      getModels: (props) => this.getModels(props),
+      getSubCategories: (props) => this.getSubCategories(props),
+      getSubModels: (props) => this.getSubModels(props),
+      hasSubModel: async (props) => this._props.idsCache.hasSubModel(props),
+    };
+    this._visibilityHelper = new ModelsTreeVisibilityHelper({
+      viewport: this._props.viewport,
+      idsCache: this._props.idsCache,
+      alwaysAndNeverDrawnElementInfo: this._props.alwaysAndNeverDrawnElementInfo,
+      overrideHandler: this._props.overrideHandler,
+      baseIdsCache,
+      overrides: this._props.overrides,
+    });
```

Comment:
With https://github.com/iTwin/viewer-components-react/issues/1421 we'll be able to just pass `this._props.idsCache` straight to `ModelsTreeVisibilityHelper` without the need to also pass `baseIdsCache`, right? If yes, please add a short comment with a link to #1421.